Split resume data management to own module, add periodic saving

gazpachoking opened the pull request:
Not really sure if this one is ready yet, just opening a PR to get some thoughts and track it.

gazpachoking committed:
Save resume data every 60 seconds.

aresch commented on spritzle/core.py:
```diff
@@ -105,17 +106,25 @@ def __init__(self, config, state_dir=None):
 
+    async def _save_resume_data_loop(self):
+        while True:
+            await asyncio.sleep(60)
```
We should make this a config value.

aresch reviewed (changes requested):
I was also thinking about moving the resume data stuff into a separate file/class (ResumeData) to reduce some of the clutter building up in Core. What do you think?

gazpachoking commented:
Yeah, I was also wondering about moving it, after I had to add all those members. Seems good. Making that time value configurable also seems good, will do.

gazpachoking committed:
Merge remote-tracking branch 'upstream/master' into save_resume_data_loop

gazpachoking committed:
Move resume data management to own module

gazpachoking commented:
The `cancel` call to stop the resume data manager causes an uncaught `CancelledError`, which I'm not sure how to properly catch yet.

gazpachoking committed:
Catch the CancelledError cleanly when stopping resume data loop

gazpachoking commented:
Okay, cancelling works cleanly now.

gazpachoking committed:
Prevent interrupting the resume data save process during shutdown

aresch commented on spritzle/resume_data.py:
```diff
@@ -0,0 +1,127 @@
+        self.core = core
+        self.loop = asyncio.get_event_loop()
+        self.save_loop_task = None
+        self.run = False
```
I don't see where this is used.

aresch commented on spritzle/resume_data.py:
```diff
@@ -0,0 +1,127 @@
+            await self.save_loop_task
+        log.debug("Resume data manager stopped.")
+
+    async def save_loop_coro(self):
```
Rename this to 'save_loop' the _coro is redundant.

aresch commented on spritzle/core.py:
```diff
@@ -104,13 +95,15 @@ def __init__(self, config, state_dir=None):
         await self.load_session_state()
         await self.alert.start(self.session)
-        await self.load_resume_data()
+        await self.resume_data.load()
```
I think we could probably just have resume_data.start() call load() unless there is some scenario where we would start it without first loading the resume data.

aresch commented on spritzle/core.py:
```diff
@@ -104,13 +95,15 @@ def __init__(self, config, state_dir=None):
         log.debug('Core stopping..')
         await self.save_session_state()
-        await self.save_resume_data()
+        await self.resume_data.stop()
```
Likewise as above, I think we could have resume_data.stop() call save().

aresch reviewed (changes requested):
I think this mostly looks good but we should probably be adding tests for some of this.. I should have done that when I originally wrote this code. Sorry.

gazpachoking committed:
Cleanups of resume_data module

gazpachoking committed:
Pause libtorrent session before final resume data save

gazpachoking commented:
I agree with all those notes and made appropriate cleanups. I also added a pause to the libtorrent session before the resume data stop gets called. As for tests, I agree, I'll get some in here. You good with the [test changes branch](#9)? I'd prefer to build on top of those if they are good.

gazpachoking committed:
Merge remote-tracking branch 'upstream/master' into save_resume_data_loop

gazpachoking committed:
Ensure lt test settings are applied outside aiohttp app context

gazpachoking committed:
Ensure test core is stopped when outside aiohttp app context

gazpachoking committed:
Fix loading of resume data on resume data manager start

gazpachoking committed:
Add some rudimentary resume data tests

gazpachoking commented:
Okay, I've added a few rudimentary tests for the resume data. I was having a bit of trouble figuring out hot to test things more than superficially.  Let me know if you can think of some other good things to test.

Also put a couple of improvements into the testing framework that were lost with my last improvement.
- If you use `core` fixture outside of the aiohttp app context, and don't stop it, it will be stopped automatically. (but not started automatically, it's nice to be able to change settings/mock functions before that happens)
- If you use `core` fixture outside of aiohttp app context, the default libtorrent settings for tests still get applied (unless they are overridden by the specific test.)

gazpachoking commented:
Here are some things I'm wondering about:
- Is the test for the save timer too fragile? Could the time it takes to do other things mess that up?
- Should the folder with the test torrents be under the `resource` directory in the tests folder, or just the tests root? (I put the resume data folder there too, but that one seems even less appropriate.)
- Should we have the `torrent_dir` (and `resume_data_dir`) for tests be importable, instead of defined in individual test files? (Maybe in `tests/__init__` such that you can `from spritzle.tests import torrent_dir`)
- I actually started to doubt the automatic core stopping mechanism I just added to `core` fixture. Does that help anything? It might cover up problems if we have any tests designed to see if the core is stopping gracefully via normal mechanisms. (EDIT: On second (third) thought, we would be testing that before the fixture cleanup actually ran.)

gazpachoking committed:
Make file open call a bit more normal

gazpachoking committed:
Standardize and shorten resume data loop test a bit

aresch commented on spritzle/tests/conftest.py:
```diff
@@ -34,9 +29,19 @@ def app(core):
         'alert_mask': 0,
         'stop_tracker_timeout': 0,
     }
+    with patch.object(core, 'start', partial(core.start, settings=settings)):
```
Why do you need to patch this? Can't you just do:
```
loop.run_until_complete(core.start(settings=settings))
```

gazpachoking replied:
I wanted the core to be able to be started from the app, or from a test, and still get those default settings.

aresch commented on spritzle/tests/test_resume_data.py:
```diff
@@ -0,0 +1,69 @@
+torrent_dir = Path(__file__).resolve().parent / 'resource' / 'torrents'
+
+
+async def test_load(core):
```
Could we create a different fixture for these tests if we don't want the core to be started or stopped by the fixture?

aresch replied:
Something like this.. we'd probably have to move the settings out to a module level attribute so we can pass it in when call core.start() in tests using core_nostart (not sold on this name) fixture. Maybe we should make Core take a settings dict in it's __init__ instead of start().

```python
 def core_init():
     config = Config(in_memory=True, config_dir='/tmp')
     state_dir = Path(tempfile.mkdtemp(prefix='spritzle-test'))
     return Core(config, state_dir)

@pytest.fixture(scope='function')
def core_nostart(loop):
    yield core_init()

@pytest.fixture(scope='function')
def core(loop):
    core = core_init()
    settings = {
        'enable_upnp': False,
        'enable_natpmp': False,
        'enable_lsd': False,
        'enable_dht': False,
        'anonymous_mode': True,
        'alert_mask': 0,
        'stop_tracker_timeout': 0,
    }
    with patch.object(core, 'start', partial(core.start, settings=settings)):
        yield core
    if core.session is not None:
        loop.run_until_complete(core.stop())
    shutil.rmtree(str(state_dir))
```

gazpachoking replied:
If stopping is important, I think it might be okay the way it is. The stopping will happen if you forget/something goes wrong. It won't start automatically, but if you are using `app` (and starting it), or `cli` fixtures it will start and stop automatically. If you aren't using those other fixtures, chances are you wanted to control that manually anyway.

aresch replied:
Oh right, because we're using the app it calls the on_shutdown handler which calls core.stop()?

aresch replied:
I still don't quite understand why it's ok to do this though because in your tests you are calling core.start() after it's already started.. this is something that you shouldn't do as I don't think I ever intended start() to be idempotent and it could have strange side-effects in the future.

We should only call start() and stop() once per test.

aresch commented on spritzle/tests/test_resume_data.py:
```diff
@@ -0,0 +1,69 @@
+    assert len(list(core.state_dir.iterdir())) == 0
+    with open(torrent_dir / 'random_one_file.torrent', mode='rb') as f:
+        torrent_info = lt.torrent_info(lt.bdecode(f.read()))
+    core.session.add_torrent({'ti': torrent_info})
```
Would it make more sense to use our API to add the torrent? I worry that if we add more state into the POST /torrent path we may miss something here and break the test. Also, if libtorrent API changes its another spot we need to update.

gazpachoking replied:
I was wondering about that... I was thinking it ends up testing more than just the resume data, but I'm fine with that. I'll switch it.

aresch commented on spritzle/tests/test_resume_data.py:
```diff
@@ -0,0 +1,69 @@
+    assert len(list(core.state_dir.iterdir())) == 1
+
+
+async def test_resume_data_save_loop(core):
```
Yea, this test is a bit concerning since it could be flaky due to timing but I'm not sure how else to test that a timed loop is functioning.

aresch commented on spritzle/tests/test_resume_data.py:
```diff
@@ -0,0 +1,69 @@
+    with asynctest.patch('spritzle.resume_data.ResumeData.save') as mock_save:
+        await core.start()
+        await asyncio.sleep(0.41)
+        assert mock_save.call_count == 4
```
Maybe to reduce the chance of flakiness we could just check we're greater than some value?

int((sleep_time / save_frequency) - 1)

```
assert mock_save.call_count > 3
```

gazpachoking replied:
Maybe I'll just allow it to be off by one, and make sure the 2 tested results are expected to be more than 2 apart.

aresch commented on spritzle/resume_data.py:
```diff
@@ -0,0 +1,128 @@
+            if save_fut and not save_fut.done():
+                await save_fut
+
+    async def on_save_resume_data_alert(self, alert):
```
We could test these by:
* adding the torrent to the session
* calling this function with a mock alert
* assert that the file was created with some data in it, maybe bdecode it and look for a key we know should be there

gazpachoking replied:
Isn't that sort of already happening in the 'test_save' test? (except instead of a mocked alert, a real one comes in.) I agree on adding a bit to bdecode the data and test it's like what we expect though.

aresch reviewed (changes requested):
* We can move the test torrents to a different directory, that's fine. I don't think they need to be in resource, it just happened that way.
* Yes, we could add the directories to a common location that is importable (__init__.py or common.py would be fine).
* I think stopping the core is important for reasons I can't quite remember ;) I had a lot of trouble with the tests causing segfaults because there were too many libtorrent sessions running and alert manager loops polling. I don't think we should get rid of it unless we absolutely need to. I added some comments on maybe adding another fixture to handle cases where we want to control the starting and stopping of the core in our tests.

gazpachoking committed:
Allow resume data loop test to be off by one

gazpachoking committed:
Make the resume data save test actually test the resume data is saved
(also fix so the resume data does get saved)

gazpachoking committed:
Move test torrents and resume data directly under tests folder
Make their paths importable from spritzle.tests

aresch commented on spritzle/resume_data.py:
```diff
@@ -78,7 +78,7 @@ def __init__(self, core):
         p = Path(self.core.state_dir, alert.torrent_name + '.resume')
         r = lt.write_resume_data(alert.params)
-        r.update(self.core.torrent_data[info_hash])
+        r.update(self.core.torrent_data.get(info_hash, {}))
```
I don't think we should change this because torrent_data for this torrent should exist when saving resume data or there is a bug somewhere.

gazpachoking replied:
Yeah, this is because I used the libtorrent add and not the POST handler turns out.

aresch commented on spritzle/tests/test_resume_data.py:
```diff
@@ -47,8 +47,13 @@
     with open(torrent_dir / 'random_one_file.torrent', mode='rb') as f:
         torrent_info = lt.torrent_info(lt.bdecode(f.read()))
     core.session.add_torrent({'ti': torrent_info})
+    # We need time for the torrent added alert to come in to initiate save
```
I wonder if it would better to add an alert manager handler to know exactly when the resume data has been saved instead of sleeping. We could probably wait for save_resume_data_alert before continuing.

gazpachoking committed:
Save new torrent resume data on add (don't rely on alert)
Add resume_data.save_torrent for single torrent resume save, with completion future

aresch commented:
Ok, I think this looks good now. Thanks for doing this.

gazpachoking commented:
No prob, thanks for all the good feedback. Will probably need some tweaking, but I think it's good progress, and more tested.
